### PR TITLE
chore: add integration tests for CORS

### LINKED ISSUE #1984: chore: add tests for Tailcall CORS
If the request comes in for `tailcall.run` we automatically enable CORS for those request. Write an integration test for it. 

Get started here — https://tailcall.run/developers/integration-testing/

### PR DESCRIPTION
**Pull request by varshith257**
**Summary:**  

1. The set_common_headers function has been created to handle common header settings to remove code duplication patterns.
2. Added Integration tests for CORS .

**Issue Reference(s):**  
- Fixes #1984 

**Build & Testing:**

- [x] I ran `cargo test` successfully.
- [x] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [x] I have added relevant unit & integration tests.
- [x] I have updated the [documentation] accordingly.
- [x] I have performed a self-review of my code.
- [x] PR follows the naming convention of `<type>(<optional scope>): <title>`

[documentation]: https://github.com/tailcallhq/tailcallhq.github.io/tree/develop/docs

### CODE REVIEW

**Comment by varshith257** [comment]
@tusharmath I think it's ready for your review. If any suggestions or modifications needed, I am ready to incorporate them

File: tests/core/cors.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+            .unwrap();
+
+        let response = handle_origin_tailcall::<GraphQLRequestLike>(req, app_ctx, &mut req_counter).await.unwrap();
+
+        // It checks the CORS headers in the response
+        assert_eq!(response.status(), StatusCode::OK);
+        assert_eq!(response.headers().get(ACCESS_CONTROL_ALLOW_ORIGIN), Some(&HeaderValue::from_static("https://tailcall.run")));
+        assert_eq!(response.headers().get(ACCESS_CONTROL_ALLOW_METHODS), Some(&HeaderValue::from_static("GET, POST, OPTIONS")));
+        assert_eq!(response.headers().get(ACCESS_CONTROL_ALLOW_HEADERS), Some(&HeaderValue::from_static("*")));
+    }
```
Comment (tusharmath) [inline tests/core/cors.rs:27]: Can we write an integration test like this — https://tailcall.run/developers/integration-testing/

**Comment by varshith257** [comment]
Any suggestions on why tests are failing?

**Comment by tusharmath** [comment]
You had to run `cargo test` locally to generate the snapshots.

File: src/core/http/request_handler.rs [hunk 2]
```diff
@@ -73,25 +73,28 @@ fn update_cache_control_header(
...
+    req_ctx.extend_x_headers(headers);
+}
+
+pub fn update_response_headers(
+    resp: &mut hyper::Response<hyper::Body>,
+    req_ctx: &RequestContext,
+    app_ctx: &AppContext,
+) {
+    set_common_headers(resp.headers_mut(), app_ctx, req_ctx);
 }
```
Comment (tusharmath) [inline src/core/http/request_handler.rs:98]: These changes doesn't seem to be doing much. Can we revert it to the one it was in main?

**Comment by varshith257** [comment]
@tusharmath Any tip for trying to get optimise the request_handler.rs file from a long ?

**Review by tusharmath — APPROVED** [review]

File: tests/execution/test-cors-tailcall-run.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+- method: OPTIONS
+  url: http://localhost:8000/graphql
+  headers:
+    Origin: https://tailcall.run
+  expected:
+    status: 200
+    headers:
+      access-control-allow-origin: https://tailcall.run
+      access-control-allow-methods: GET, POST, OPTIONS
+      access-control-allow-headers: "*"
```
Comment (tusharmath) [inline tests/execution/test-cors-tailcall-run.md:31]: Let's add a few more tests, with different methods, origin, and header settings. To ensure that our logic is working properly.

File: tests/execution/test-cors-tailcall-run.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+type Query {
+  example: String @http(path: "/example")
+}
+```
+
+```yml @test
+# CORS test
+- method: OPTIONS
```
Comment (tusharmath) [inline tests/execution/test-cors-tailcall-run.md:22]: If cors is enabled we need to try to make requests with `GET` and `POST` methods also.

**Review by tusharmath — CHANGES_REQUESTED** [review]
Fix review comments. Whenever completed, please mark it as ready to review.

**Comment by varshith257** [comment]
I am unable to run `cargo test` to generate snapshots. This emitting lot of errors

File: tests/execution/test-cors-tailcall-run.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
+# Verify the correct CORS headers for requests from https://tailcall.run
+
+```graphql @config
+schema
+  @upstream(baseURL: "http://example.com")
+  @server(
+    port: 8000
+    headers: {
+      cors: {allowHeaders: ["*"], allowMethods: ["GET", "POST", "OPTIONS"], allowOrigins: ["https://tailcall.run"]}
```
Comment (meskill) [inline tests/execution/test-cors-tailcall-run.md:9]: The issue is intended to test headers that put automatically for `tailcall.run` and therefore we don't need to specify headers in test - we need to test that it works without headers

File: tests/execution/test-cors-tailcall-run.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    }
+  ) {
+  query: Query
+}
+
+type Query {
+  example: String @http(path: "/example")
```
Comment (meskill) [inline tests/execution/test-cors-tailcall-run.md:16]: need to mock this request

File: tests/execution/test-cors-tailcall-run.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+# Additional CORS test for GET request with custom headers
+- method: GET
+  url: http://localhost:8000/graphql
+  headers:
+    Origin: https://tailcall.run
+    X-Custom-Header: value
+  expected:
+    status: 200
+    headers:
+      access-control-allow-origin: https://tailcall.run
```
Comment (meskill) [inline tests/execution/test-cors-tailcall-run.md:109]: we use snapshots and not explicit `expected` definitions in tests. That should be removed and proper snapshots should be generated

**Review by meskill — CHANGES_REQUESTED** [review]

**Comment by meskill** [comment]
Closing for now since it requires additional work. Feel free to reopen this when it's ready. If you have any troubles with code/tooling please ask in our #contributors channel on Discord

**Pull request closed without merging**